cmd/syncthing/cli/main.go

```diff
@@ -8,165 +8,83 @@ package cli
-	// This is so that the usage text is the same
-	fakeFlags := []cli.Flag{
-		cli.StringFlag{
-			Name:  "gui-address",
-			Usage: "Override GUI address to `URL` (e.g. \"192.0.2.42:8443\")",
-		},
-		cli.StringFlag{
-			Name:  "gui-apikey",
-			Usage: "Override GUI API key to `API-KEY`",
-		},
-		cli.StringFlag{
-			Name:  "home",
-			Usage: "Set configuration and data directory to `PATH`",
-		},
-		cli.StringFlag{
-			Name:  "config",
-			Usage: "Set configuration directory (config and keys) to `PATH`",
-		},
-		cli.StringFlag{
-			Name:  "data",
-			Usage: "Set data directory (database and logs) to `PATH`",
-		},
-	}
-
-	// Construct the actual CLI
-	app := cli.NewApp()
-	app.Author = "The Syncthing Authors"
-	app.Metadata = map[string]interface{}{
-		"clientFactory": clientFactory,
-	}
-	app.Commands = []cli.Command{{
-		Name:  "cli",
-		Usage: "Syncthing command line interface",
-		Flags: fakeFlags,
-		Subcommands: []cli.Command{
-			configCommand,
-			showCommand,
-			operationCommand,
-			errorsCommand,
-			debugCommand,
-			{
-				Name:     "-",
-				HideHelp: true,
-				Usage:    "Read commands from stdin",
-				Action: func(ctx *cli.Context) error {
-					if ctx.NArg() > 0 {
-						return errors.New("command does not expect any arguments")
-					}
-
-					// Drop the `-` not to recurse into self.
-					args := make([]string, len(cliArgs)-1)
-					copy(args, cliArgs)
-
-					fmt.Println("Reading commands from stdin...", args)
-					scanner := bufio.NewScanner(os.Stdin)
-					for scanner.Scan() {
-						input, err := shlex.Split(scanner.Text())
-						if err != nil {
-							return fmt.Errorf("parsing input: %w", err)
-						}
-						if len(input) == 0 {
-							continue
-						}
-						err = app.Run(append(args, input...))
-						if err != nil {
-							return err
-						}
-					}
-					return scanner.Err()
-				},
-			},
-		},
-	}}
-
-	return app.Run(cliArgs)
+	kongCtx.Bind(context)
+	return nil
 }
 
-func parseFlags(c *preCli) error {
-	// kong only needs to parse the global arguments after "cli" and before the
-	// subcommand (if any).
-	if len(os.Args) <= 2 {
-		return nil
-	}
-	return parseFlagsWithArgs(os.Args[2:], c)
-}
+type stdinCommand struct{}
+
+func (*stdinCommand) Run() error {
+	// Drop the `-` not to recurse into self.
+	args := make([]string, len(os.Args)-1)
+	copy(args, os.Args)
 
-func parseFlagsWithArgs(args []string, c *preCli) error {
-	for i := 0; i < len(args); i++ {
-		if !strings.HasPrefix(args[i], "--") {
-			args = args[:i]
-			break
+	fmt.Println("Reading commands from stdin...", args)
+	scanner := bufio.NewScanner(os.Stdin)
+	for scanner.Scan() {
+		input, err := shlex.Split(scanner.Text())
+		if err != nil {
+			return fmt.Errorf("parsing input: %w", err)
 		}
-		if !strings.Contains(args[i], "=") {
-			i++
+		if len(input) == 0 {
+			continue
+		}
+		cmd := exec.Command(os.Args[0], append(args[1:], input...)...)
+		out, err := cmd.CombinedOutput()
+		fmt.Print(string(out))
```

Comment:
This works quite differently however, we now have subprocesses instead of the execution happening in the same process.
Could we not get kong to re-parse the arguments?

Reply:
I missed this, it's not super pretty. I wonder if there's a compelling use case here to begin with... `#!/bin/syncthing cli` scripts?

Reply:
The original intent (I think), which is now broken, was that you do multiple actions with a single roundtrip.

Namely, it would fetch the config at startup, run series of your commands from a text file/stdin, applying all the config changes to the config object in memory, before posting the config back. This way you have one round trip, one set of locking and file renaming shenanigans for a large set of  changes.